### PR TITLE
upgrade space-ros to jazzy/noble (closes #157)

### LINKED ISSUE #157: Support Jazzy Jalisco
ROS 2 Jazzy will be released by the end of May. We need to discuss what we want to do, I think the options are:
1. Ignore it for this release (2024.7.0) - release humble only
2. Move main to be based on jazzy and release only jazzy going forward
3. Branch humble to it's own branch, move main to jazzy and release jazzy, continuing to update / release humble when there's a new sync release, but using main (jazzy) for new development

### PR DESCRIPTION
**Pull request by asimonov**
upgrade `space-ros` to Jazzy/Noble (ubuntu24)

current status:
* [x] cobra fix merged (https://github.com/nimble-code/Cobra/pull/68). 
* [x] new Cobra release (https://github.com/nimble-code/Cobra/issues/71)
* [x] change ament_cobra to reference new Cobra 4.8 release (https://github.com/ament/ament_cobra/pull/21)
* [x] ament_cobra release 0.12.0 (https://github.com/ament/ament_cobra/issues/23)
* [x] ikos PR to fix Ubuntu22 build (https://github.com/NASA-SW-VnV/ikos/pull/274)
* [x] need new ikos release (https://github.com/NASA-SW-VnV/ikos/issues/280)
* [x] disable tests for ament_lint as it gives error
* [x] fast-forward `jazzy` branch to `main`. there is no history on jazzy, so fast forward is ok
* [ ] optional: backport unittest fix to ament_lint/spaceros branch (discussion #218, https://github.com/ament/ament_lint/pull/507)
* [ ] add image push from space-ros jazzy branch to github registry or docker hub - to start testing downstream repos

### CODE REVIEW

**Comment by EzraBrooks** [comment]
I think we probably need to talk about branching strategy, because I don't think we have the engineering hours to do all of the complex networking of back porting and such that the ROS upstream does.

**Comment by asimonov** [comment]
> @asimonov - can you change the target branch to `jazzy`?

done. sorry for delay.

@EzraBrooks we did a quick discussion with Matt and Ivan during the call last thursday. 

Jazzy upgrade is likely to break a lot of things in other space-ros repos. And fixing those may take more time than we have before release in October. So we thought it would be good to start this work on jazzy branch of main repo. And probably introduce jazzy branches in other repos to develop the fixes. We may also need jazzy image tags on docker registry (ghcr.io?) to not keep rebuilding them from scratch.

What to do with the branches after jazzy works is open for discussion. One alternative is to merge jazzy to main, but branch (or tag?) humble for possible later fixes. Do we want/have capacity to maintain humble and jazzy? May be not... But we may have to do that for the duration of whole project upgrade.

As for when to merge those branches - Matt and Ivan slightly disagreed. Matt is in favour of releasing space-ros on jazzy, but keeping demos on humble for some time. Ivan says we should release all repos on same version...

**Comment by asimonov** [comment]
what shall we do about cobra? I created issue on ament_cobra repo (https://github.com/ament/ament_cobra/issues/18) but looks like it only has Michael and Steven as maintainers. Not sure they have capacity to look at it...

does anyone have understanding of cobra/ament funcationality that we have?

**Comment by asimonov** [comment]
> what shall we do about cobra? I created issue on ament_cobra repo ([ament/ament_cobra#18](https://github.com/ament/ament_cobra/issues/18)) but looks like it only has Michael and Steven as maintainers. Not sure they have capacity to look at it...
> 
> does anyone have understanding of cobra/ament funcationality that we have?

ok, there is a PR in Cobra to fix the build failure: https://github.com/nimble-code/Cobra/pull/68/files

I am not sure how it works for Cobra releases. probably the bin_linux directory is there already. but for new checkouts it needs to be created, as per PR.

**Comment by asimonov** [comment]
@Bckempa can you trigger the build again? I fixed cobra build (as a workaround now, until cobra makes a new release)

so, here everything we used to have should build/test

**Comment by asimonov** [comment]
I see locally that the build works, but testing fails on ament_lint. I am not sure what is wrong, but I started to investigate what is ament_lint and I see we have other work in space-ros (https://github.com/space-ros/space-ros/issues/57) around ament_lint.

interestingly we use `spaceros` branch of ament_lint (https://github.com/ament/ament_lint/tree/spaceros) which is 101 commits behind `rolling`.

so, the question is - do we want to more that branch forward to match jazzy?

**Comment by asimonov** [comment]
the build-testing does not work on ament_lint package, gives the following error:

```
      +build-testing | [1.656s] WARNING:colcon.colcon_core.verb.test:Ignored '--retest-until-pass' for package 'ament_lint' since 'unittest' does not support the usage
      +build-testing | --- stderr: ament_lint

      +build-testing | ----------------------------------------------------------------------
      +build-testing | Ran 0 tests in 0.000s

      +build-testing | NO TESTS RAN
      +build-testing | ---

      +build-testing | Failed   <<< ament_lint [1.14s, exited with code 5]
```

need to investigate

**Comment by asimonov** [comment]
not sure i have done the rebase right... may be we should have rebased jazzy branch on main first?

**Comment by asimonov** [comment]
> not sure i have done the rebase right... may be we should have rebased jazzy branch on main first?

@mkhansenbot can you re-base jazzy on main by any chance?

**Comment by EzraBrooks** [comment]
I believe per previous discussions we are not rewriting history on long-running branches in this project - we should just merge main into jazzy instead of rebasing (which you can open a PR to do)

**Comment by asimonov** [comment]
> I believe per previous discussions we are not rewriting history on long-running branches in this project - we should just merge main into jazzy instead of rebasing (which you can open a PR to do). rebasing feature branches like this one, obviously, is fine - we just don't want to muck around with history on mainline branches

@EzraBrooks there is nothing on jazzy yet, just one PR pending. So the effect will be the same

File: Earthfile [hunk 1]
```diff
@@ -214,38 +219,86 @@ rosdep:
...
   # TODO: Consider pushing pre-built dev images to the registry.
   # SAVE IMAGE --push osrf/space-ros-dev:latest osrf/space-ros-dev:$tag
 
 build-testing:
   FROM +build-dev
+  # exclude ament_lint tests as they give error on Jazzy
   RUN . install/setup.sh && \
       colcon test \
         --retest-until-pass 2 \
+        --packages-skip ament_lint \
```
Comment (asimonov) [inline Earthfile:301]: for now remove ament_lint itself from testing as it gives an error. without it we produce the testing artifacts

Comment (asimonov) [reply Earthfile:301]: this version does not fail on tests and produces some artifacts that we can compare to humble.

but the issue of synching spaceros and jazzy branches of ament_lint still remains...

**Comment by asimonov** [comment]
@Bckempa can you kick off the build?

**Comment by asimonov** [comment]
cobra 4.8 is released, but there are issues with building it. 

there are lots of changes in makefile: https://github.com/nimble-code/Cobra/issues/71#issuecomment-2428871636

investigating...

File: Earthfile (hunk 1 above)
Comment (xfiderek) [reply Earthfile:301]: Please take a look at this comment @asimonov https://github.com/space-ros/space-ros/issues/218#issuecomment-2429522914

**Comment by asimonov** [comment]
> Just started the build

@mkhansenbot can you re-run the build?

I have fixed https://github.com/ament/ament_cobra/pull/21 - it now works on my local setup. The build/test should now succeed and generate test outputs that we can compare with humble

**Comment by asimonov** [comment]
> It looks like it passed except for the commit message check

@mkhansenbot thanks for triggering!

I created PR from the latest state of 'main', so some commits that are shown in this PR are actually old ones. can you fast forward this 'jazzy' branch to match current 'main'?

**Comment by asimonov** [comment]
it builds and runs tests ok under my cloned repo: https://github.com/asimonov/space-ros/actions/runs/11569301735

can someone trigger build here please?

**Comment by asimonov** [comment]
@xfiderek @Bckempa @mkhansenbot @EzraBrooks 

all done, apart from tag for ament_cobra. Waiting for Ivan.

can someone review?

can someone sync jazzy branch to main?

**Comment by EzraBrooks** [comment]
just hit the run button on CI

**Comment by EzraBrooks** [comment]
I've just overwritten the `jazzy` branch with the contents of the `main` branch since it was 0 commits ahead and thus shouldn't require a merge commit for traceability.

**Comment by asimonov** [comment]
> I've just overwritten the `jazzy` branch with the contents of the `main` branch since it was 0 commits ahead and thus shouldn't require a merge commit for traceability.

Thanks!

PR still shows old commits as part of it, i will investigate

**Comment by EzraBrooks** [comment]
the fork point of your branch (AKA `157_jazzy_upgrade^`) is probably still pointing at the old location on `jazzy`, you probably just need to rebase again

**Comment by asimonov** [comment]
> the parent commit of your branch (AKA `157_jazzy_upgrade^`) is probably still pointing at the old location on `jazzy`, you probably just need to rebase again

not sure... I have checked/synced all remotes/branches locally with github. the 157_jazzy_upgrade correctly shows 5 commits ahead of upstream on github. But the PR still shows historic commits. I tried pushing a dummy commit hoping github to figure out the history changed, but it still counts the previous commits as this PR...

I feel something needs to be changed for PR, but not sure what/how...

or better to create new PR?

**Comment by asimonov** [comment]
> You might be able to recreate the MERGE_HEAD by changing the base branch to main and then changing it back to jazzy. this is a little odd

I have just pushed the commit to fix ament_cobra version. And somehow now PR shows good commit history.

Can we kick off CI again please?

File: Earthfile [hunk 2]
```diff
@@ -171,13 +170,16 @@ rosdep:
...
+      rosdep install -s -y \
         --from-paths src --ignore-src \
         --rosdistro ${ROS_DISTRO} \
         # `urdfdom_headers` is cloned from source, however rosdep can't find it.
         # It is because package.xml manifest is missing. See: https://github.com/ros/urdfdom_headers
         # Additionally, IKOS must be excluded as per: https://github.com/space-ros/docker/issues/99
-        --skip-keys "$(tr '\n' ' ' < 'excluded-pkgs.txt') urdfdom_headers ikos"
+        --skip-keys "$(tr '\n' ' ' < 'excluded-pkgs.txt') urdfdom_headers ikos" > rosdep-commands.sh && \
+      chmod u+x rosdep-commands.sh && \
+      ./rosdep-commands.sh
```
Comment (EzraBrooks) [inline Earthfile:181]: If we want to cache rosdep better (in a later PR), we could end this at the `> rosdep-commands.sh` and then `COPY` that file into another layer and install there - then the install wouldn't be re-run if the script didn't change.

**Review by EzraBrooks — APPROVED** [review]
this LGTM with one minor comment about future work

**Comment by asimonov** [comment]
I did not hear from Ivan on merging Blazej change to ament_lint (https://github.com/ament/ament_lint/pull/507).

I suggest we dont wait for that, as it is optional/nice to have fix.

I think we can merge and proceed migrating all the demos to Jazzy so we can release Jazzy by next milestone. There may be a lot of work there...

**Pull request merged**